File: Source/jli_game_engine/version-1.0/doc/doxygen/njliStubs/SceneStateMachine.lua

--Test the functions for SceneStateMachine
local Test = {}

local function SceneStateMachine_calculateSerializeBufferSize_()
--//! [SceneStateMachine_calculateSerializeBufferSize_]
-- @TODO: Test: SceneStateMachine:calculateSerializeBufferSize()
--//! [SceneStateMachine_calculateSerializeBufferSize_]

end

local function SceneStateMachine_serialize_btSerializer()
--//! [SceneStateMachine_serialize_btSerializer]
-- @TODO: Test: SceneStateMachine:serialize(btSerializer)
--//! [SceneStateMachine_serialize_btSerializer]

end

local function SceneStateMachine_getClassName_()
--//! [SceneStateMachine_getClassName_]
-- @TODO: Test: SceneStateMachine:getClassName()
--//! [SceneStateMachine_getClassName_]

end

local function SceneStateMachine_getType_()
--//! [SceneStateMachine_getType_]
-- @TODO: Test: SceneStateMachine:getType()
--//! [SceneStateMachine_getType_]

end

local function SceneStateMachine_createArray_size()
--//! [SceneStateMachine_createArray_size]
-- @TODO:Test: NJLI.SceneStateMachine.createArray(size)
--//! [SceneStateMachine_createArray_size]

end

local function SceneStateMachine_destroyArray_array()
--//! [SceneStateMachine_destroyArray_array]
-- @TODO:Test: NJLI.SceneStateMachine.destroyArray(array)
--//! [SceneStateMachine_destroyArray_array]

end

local function SceneStateMachine_create_()
--//! [SceneStateMachine_create_]
-- @TODO:Test: NJLI.SceneStateMachine.create()
--//! [SceneStateMachine_create_]

end

local function SceneStateMachine_create_builder()
--//! [SceneStateMachine_create_builder]
-- @TODO:Test: NJLI.SceneStateMachine.create(builder)
--//! [SceneStateMachine_create_builder]

end

local function SceneStateMachine_clone_object()
--//! [SceneStateMachine_clone_object]
-- @TODO:Test: NJLI.SceneStateMachine.clone(object)
--//! [SceneStateMachine_clone_object]

end

local function SceneStateMachine_copy_object()
--//! [SceneStateMachine_copy_object]
-- @TODO:Test: NJLI.SceneStateMachine.copy(object)
--//! [SceneStateMachine_copy_object]

end

local function SceneStateMachine_destroy_object()
--//! [SceneStateMachine_destroy_object]
-- @TODO:Test: NJLI.SceneStateMachine.destroy(object)
--//! [SceneStateMachine_destroy_object]

end

local function SceneStateMachine_load_object_L_stack_index()
--//! [SceneStateMachine_load_object_L_stack_index]
-- @TODO:Test: NJLI.SceneStateMachine.load(object, L, stack_index)
--//! [SceneStateMachine_load_object_L_stack_index]

end

local function SceneStateMachine_type_()
--//! [SceneStateMachine_type_]
-- @TODO:Test: NJLI.SceneStateMachine.type()
--//! [SceneStateMachine_type_]

end


function Test:testAll()
	SceneStateMachine_calculateSerializeBufferSize_()
	SceneStateMachine_serialize_btSerializer()
	SceneStateMachine_getClassName_()
	SceneStateMachine_getType_()
	SceneStateMachine_createArray_size()
	SceneStateMachine_destroyArray_array()
	SceneStateMachine_create_()
	SceneStateMachine_create_builder()
	SceneStateMachine_clone_object()
	SceneStateMachine_copy_object()
	SceneStateMachine_destroy_object()
	SceneStateMachine_load_object_L_stack_index()
	SceneStateMachine_type_()
end

return Test

File: Source/jli_game_engine/version-1.0/src/SceneStateMachine.cpp
//
//  SceneStateMachine.cpp
//  JLIGameEngineTest
//
//  Created by James Folk on 12/6/14.
//  Copyright (c) 2014 James Folk. All rights reserved.
//

#include "SceneStateMachine.h"
#include "JLIFactoryTypes.h"
#include "World.h"
#include "SceneStateMachineBuilder.h"
#define FORMATSTRING "{\"njli::SceneStateMachine\":[]}"
#include "btPrint.h"

namespace njli
{
    SceneStateMachine::SceneStateMachine():
    AbstractStateMachine<Scene>()
    {
        
    }
    
    SceneStateMachine::SceneStateMachine(const AbstractBuilder &builder):
    AbstractStateMachine<Scene>(builder)
    {
        
    }
    
    SceneStateMachine::SceneStateMachine(const SceneStateMachine &copy):
    AbstractStateMachine<Scene>(copy)
    {
        
    }
    
    SceneStateMachine::~SceneStateMachine()
    {
        
    }
    
    SceneStateMachine &SceneStateMachine::operator=(const SceneStateMachine &rhs)
    {
        if(this != &rhs)
        {
            
        }
        return *this;
    }
    
    s32	SceneStateMachine::calculateSerializeBufferSize() const
    {
        //TODO: calculateSerializeBufferSize
        return 0;
    }
    
    void SceneStateMachine::serialize(void* dataBuffer, btSerializer* serializer) const
    {
        //TODO: serialize
    }
    
    const char *SceneStateMachine::getClassName()const
    {
        return "SceneStateMachine";
    }
    
    s32 SceneStateMachine::getType()const
    {
        return SceneStateMachine::type();
    }
    
    SceneStateMachine::operator std::string() const
    {
        return njli::JsonJLI::parse(string_format("%s", FORMATSTRING).c_str());
    }
    
    SceneStateMachine **SceneStateMachine::createArray(const u32 size)
    {
        return (SceneStateMachine**)World::getInstance()->getWorldFactory()->createArray(SceneStateMachine::type(), size);
    }
    
    void SceneStateMachine::destroyArray(SceneStateMachine **array, const u32 size)
    {
        World::getInstance()->getWorldFactory()->destroyArray((AbstractFactoryObject**)array, size);
    }
    
    SceneStateMachine *SceneStateMachine::create()
    {
        return dynamic_cast<SceneStateMachine*>(World::getInstance()->getWorldFactory()->create(SceneStateMachine::type()));
    }
    
    SceneStateMachine *SceneStateMachine::create(const SceneStateMachineBuilder &builder)
    {
        AbstractBuilder *b = (AbstractBuilder *)&builder;
        
        return dynamic_cast<SceneStateMachine*>(World::getInstance()->getWorldFactory()->create(*b));
    }
    
    SceneStateMachine *SceneStateMachine::clone(const SceneStateMachine &object)
    {
        return dynamic_cast<SceneStateMachine*>(World::getInstance()->getWorldFactory()->clone(object, false));
    }
    
    SceneStateMachine *SceneStateMachine::copy(const SceneStateMachine &object)
    {
        return dynamic_cast<SceneStateMachine*>(World::getInstance()->getWorldFactory()->clone(object, true));
    }
    
    void SceneStateMachine::destroy(SceneStateMachine *object)
    {
        if(object)
        {
            object->clear();
            
            World::getInstance()->getWorldFactory()->destroy(object);
        }
    }
    
    void SceneStateMachine::load(SceneStateMachine &object, lua_State *L, int index)
    {
        // Push another reference to the table on top of the stack (so we know
        // where it is, and this function can work for negative, positive and
        // pseudo indices
        lua_pushvalue(L, index);
        // stack now contains: -1 => table
        lua_pushnil(L);
        // stack now contains: -1 => nil; -2 => table
        while (lua_next(L, -2))
        {
            // stack now contains: -1 => value; -2 => key; -3 => table
            // copy the key so that lua_tostring does not modify the original
            lua_pushvalue(L, -2);
            // stack now contains: -1 => key; -2 => value; -3 => key; -4 => table
            const char *key = lua_tostring(L, -1);
            const char *value = lua_tostring(L, -2);
            if(lua_istable(L, -2))
            {
                SceneStateMachine::load(object, L, -2);
            }
            else
            {
                if(lua_isnumber(L, index))
                {
                    double number = lua_tonumber(L, index);
                    printf("%s => %f\n", key, number);
                }
                else if(lua_isstring(L, index))
                {
                    const char *v = lua_tostring(L, index);
                    printf("%s => %s\n", key, v);
                }
                else if(lua_isboolean(L, index))
                {
                    bool v = lua_toboolean(L, index);
                    printf("%s => %d\n", key, v);
                }
                else if(lua_isuserdata(L, index))
                {
//                    swig_lua_userdata *usr;
//                    swig_type_info *type;
//                    assert(lua_isuserdata(L,index));
//                    usr=(swig_lua_userdata*)lua_touserdata(L,index);  /* get data */
//                    type = usr->type;
//                    njli::AbstractFactoryObject *object = static_cast<njli::AbstractFactoryObject*>(usr->ptr);
//                    printf("%s => %d:%s\n", key, object->getType(), object->getClassName());
                    
                }
            }
            // pop value + copy of key, leaving original key
            lua_pop(L, 2);
            // stack now contains: -1 => key; -2 => table
        }
        // stack now contains: -1 => table (when lua_next returns 0 it pops the key
        // but does not push anything.)
        // Pop table
        lua_pop(L, 1);
        // Stack is now the same as it was on entry to this function
    }
    
    u32 SceneStateMachine::type()
    {
        return JLI_OBJECT_TYPE_SceneStateMachine;
    }
}